src/nvim/sign.c

```diff
@@ -77,736 +58,229 @@ static char *cmds[] = {
-/// the correct file get loaded.
-///
-/// @param buf  buffer to store sign in
-/// @param id  sign ID
-/// @param group  sign group
-static int buf_findsign(buf_T *buf, int id, char *group)
-{
-  sign_entry_T *sign;  // a sign in the signlist
+  MarkTreeIter itr[1];
+  int row = atlnum > 0 ? atlnum - 1 : 0;
+  kvec_t(MTKey) signs = KV_INITIAL_VALUE;
+  // Store and sort when removing a single sign at a specific line number.
+  if (atlnum > 0) {
+    if (!marktree_itr_get_overlap(buf->b_marktree, row, 0, itr)) {
+      return FAIL;
+    }
 
-  FOR_ALL_SIGNS_IN_BUF(buf, sign) {
-    if (sign->se_id == id && sign_in_group(sign, group)) {
-      return (int)sign->se_lnum;
+    MTPair pair;
+    while (marktree_itr_step_overlap(buf->b_marktree, itr, &pair)) {
+      if ((ns == UINT32_MAX || ns == pair.start.ns)
+          && pair.start.decor_full && decor_has_sign(pair.start.decor_full)) {
+        kv_push(signs, pair.start);
+      }
     }
+  } else {
+    marktree_itr_get(buf->b_marktree, 0, 0, itr);
   }
 
-  return 0;
-}
-
-/// Return the sign at line 'lnum' in buffer 'buf'. Returns NULL if a sign is
-/// not found at the line. If 'groupname' is NULL, searches in the global group.
-///
-/// @param buf  buffer whose sign we are searching for
-/// @param lnum  line number of sign
-/// @param groupname  sign group name
-static sign_entry_T *buf_getsign_at_line(buf_T *buf, linenr_T lnum, char *groupname)
-{
-  sign_entry_T *sign;    // a sign in the signlist
-
-  FOR_ALL_SIGNS_IN_BUF(buf, sign) {
-    if (sign->se_lnum > lnum) {
-      // Signs are sorted by line number in the buffer. No need to check
-      // for signs after the specified line number 'lnum'.
+  while (itr->x) {
+    MTKey mark = marktree_itr_current(itr);
+    if (row && mark.pos.row > row) {
       break;
     }
-
-    if (sign->se_lnum == lnum && sign_in_group(sign, groupname)) {
-      return sign;
+    if (!mt_end(mark) && mark.decor_full && decor_has_sign(mark.decor_full)
+        && (id == 0 || (int)mark.id == id)
+        && (ns == UINT32_MAX || ns == mark.ns)) {
+      if (atlnum > 0) {
+        kv_push(signs, mark);
+        marktree_itr_next(buf->b_marktree, itr);
+      } else {
+        extmark_del(buf, itr, mark, true);
+      }
+    } else {
+      marktree_itr_next(buf->b_marktree, itr);
     }
   }
 
-  return NULL;
-}
-
-/// Return the identifier of the sign at line number 'lnum' in buffer 'buf'.
-///
-/// @param buf  buffer whose sign we are searching for
-/// @param lnum  line number of sign
-/// @param groupname  sign group name
-static int buf_findsign_id(buf_T *buf, linenr_T lnum, char *groupname)
-{
-  sign_entry_T *sign;   // a sign in the signlist
-
-  sign = buf_getsign_at_line(buf, lnum, groupname);
-  if (sign != NULL) {
-    return sign->se_id;
+  if (kv_size(signs)) {
+    qsort((void *)&kv_A(signs, 0), kv_size(signs), sizeof(MTKey), sign_cmp);
+    extmark_del_id(buf, kv_A(signs, 0).ns, kv_A(signs, 0).id);
+    kv_destroy(signs);
```

Comment:
What is going on here? The docstring sounds like deleting all signs which matches some specification. If this is intentional, maybe write a small comment why the first item is picked and deleted?

Reply:
Yeah intentional. When `atlnum` is specified, only a single (sorted) sign is deleted. There is a comment some lines above:

`  // Store and sort when removing a single sign at a specific line number.`

I can add another here.